feat(api): `nvim_get_chan_info` pass 0 for current channel

nikolightsaber opened the pull request:
closes #21291

Hi first time doing this. I tried to follow the guidelines. I hope documentation and testing is good enough.

For testing I tried to follow `nvim_get_api_info()` but i did not find unit tests on this. I figured changing some of the `nvim_get_api_info()[1]` should be good enough for such a simple change.

Thanks for the feedback.

justinmk reviewed (changes requested):
Let's try to avoid adding a new api function. Why can't we do this: https://github.com/neovim/neovim/issues/21291#issuecomment-1455212995

> Instead, we could make `nvim_get_chan_info(0)` (that is, "zero" means "current channel") work.

If that doesn't work, I'm starting to think we should relax `:help api-contract` to allow `void` return values to be changed.

nikolightsaber committed:
feat(api): `nvim_get_chan_info` pass 0 for current channel

Getting current channel info was kind of annoying via RPC. Two
functions needed to be called

1. `nvim_get_api_info` which returns `[channel_id, meta_data]`
  - this results in `channel_id = api.nvim_get_api_info()[0]`
  - Here the meta_data is send but never used
2. Finally call `nvim_get_chan_info(channel_id)`

This commit reduces the need for `nvim_get_api_info` as passing 0
returns current channel info

nikolightsaber commented:
@justinmk I think i succeeded in `nvim_get_chan_info` handling 0 as an argument.

I figured that channel 0 is impossible as nvim always seem to start at 3.

I'm a bit confused on the doc and linting pipeline failing. Could it be an issue unrelated to my changes ?

The test that fail seem to be my changes as the have to to with channels. I will have a look at them if needed.

I would appreciate a review at this stage.

Thanks

zeertzjq commented on runtime/doc/news.txt:
```diff
@@ -405,6 +405,9 @@ The following changes to existing APIs or features add new behavior.
 • Attempting to set an invalid keycode option (e.g. `set t_foo=123`) no longer
   gives an error.
 
+• Allow to pass 0 to |nvim_get_chan_info| to get channel info for current
+  channel
```
```suggestion
• Passing 0 to |nvim_get_chan_info| gets channel info for the current channel.
```

zeertzjq commented on src/nvim/api/vim.c:
```diff
@@ -1633,12 +1634,16 @@ void nvim_set_client_info(uint64_t channel_id, String name, Dictionary version,
   }
+  if (chan == 0) {
+    assert(channel_id <= INT64_MAX);
+    chan = (Integer) channel_id;
```
```suggestion
    chan = (Integer)channel_id;
```

zeertzjq commented on src/nvim/api/vim.c:
```diff
@@ -1633,12 +1634,16 @@ void nvim_set_client_info(uint64_t channel_id, String name, Dictionary version,
     return (Dictionary)ARRAY_DICT_INIT;
   }
+  if (chan == 0) {
+    assert(channel_id <= INT64_MAX);
```
Internal calls will fail this assertion:
```c
/// Mask for all internal calls
#define INTERNAL_CALL_MASK (((uint64_t)1) << (sizeof(uint64_t) * 8 - 1))

/// Internal call from Vimscript code
#define VIML_INTERNAL_CALL INTERNAL_CALL_MASK

/// Internal call from Lua code
#define LUA_INTERNAL_CALL (VIML_INTERNAL_CALL + 1)
```

zeertzjq commented on src/nvim/api/vim.c:
```diff
@@ -1633,12 +1634,16 @@ void nvim_set_client_info(uint64_t channel_id, String name, Dictionary version,
   if (chan < 0) {
     return (Dictionary)ARRAY_DICT_INIT;
   }
+  if (chan == 0) {
```
```suggestion
  if (chan == 0 && !is_internal_call(channid_id)) {
```

nikolightsaber commented:
Thanks for the review @zeertzjq I implemented the changes.

zeertzjq commented on test/functional/api/vim_spec.lua:
```diff
@@ -2449,7 +2449,8 @@ describe('API', function()
 
     it('returns {} for invalid channel', function()
-      eq({}, api.nvim_get_chan_info(0))
+      -- 0 should return current channel
+      eq(testinfo, api.nvim_get_chan_info(0))
```
I think this should be moved to the `stream=stdio channel` block, before both `eq(testinfo, api.nvim_get_chan_info(1))` assertions.

justinmk reviewed (approved):
please explain what problem this is solving, in the commit message

nikolightsaber commented:
Hi @justinmk, @zeertzjq thanks for the help with this.  I enjoyed learning about the api. Quite clever with the generated code. 

I hope I can contribute some more if I find some time.
